cddagl/ui/views/main.py

```diff
@@ -2347,6 +2348,81 @@ def copy_next_soundpack(self):
 
             self.post_extraction_step3()
 
+    def preserve_custom_fonts(self):
+        """
+        Copy over any files in the previous font directory
+        that don't already exist in the current font directory.
+
+        This assumes that fonts distributed with CDDA have already
+        been extracted into the current font directory.
+        """
+        status_bar = self.get_main_window().statusBar()
+
+        join_parts = lambda parts: Path(os.path.join(*parts))
+
+        font_locations = [
+            [self.game_dir, 'font'],        # User fonts
+            [self.game_dir, 'data', 'font'] # Game fonts
+        ]
+
+        prev_font_locations = [
+            [base, 'previous_version'] + parts for base, *parts in font_locations
+        ]
+
+        # A list of tuples with the shape (CURRENT_FONT_DIR, PREV_FONT_DIR)
+        font_paths = list(tuple(
+            zip(
+                map(join_parts, font_locations),
+                map(join_parts, prev_font_locations)
+            )
+        ))
+
+        logger.info('Scanning the following user and game font directories:')
+        logger.info('{}'.format(font_locations))
+        logger.info('{}'.format(prev_font_locations))
+
+        logger.info('Mapped current to previous font dirs:')
+        logger.info('{}'.format(font_paths))
+
+        if (any(font_paths) and self.in_post_extraction):
+            logger.debug('Restoring custom fonts')
+            status_bar.showMessage(_('Restoring custom fonts'))
+
+            for font_dir, prev_font_dir in font_paths:
+                # Skip the dir if we have nothing to restore
+                if not prev_font_dir.is_dir():
+                    pass
+
+                # Create a new font directory if it doesn't already exist
+                if not font_dir.is_dir():
+                    font_dir.mkdir(exist_ok=True)
+
+                with os.scandir(prev_font_dir) as entries:
+                    previous_set = set(map(Path, entries))
+
+                with os.scandir(font_dir) as entries:
+                    current_set = set(map(Path, entries))
+
+                # Determine what font files need to be restored
+                delta = previous_set - current_set
+
+                for entry in delta:
+                    source = prev_font_dir.joinpath(entry)
+                    target =      font_dir.joinpath(entry)
+
+                    logger.debug('  Restoring {}'.format(entry))
+                    logger.debug('    Source: {}'.format(source))
+                    logger.debug('    Target: {}'.format(target))
+
+                    if os.path.isfile(source):
+                        logger.debug('    Copying file')
+                        shutil.copy2(source, target)
+                    elif os.path.isdir(source):
```

Comment:
This should probably be `if entry.is_dir():` to reuse attributes obtained from `os.scandir()`.